### PR TITLE
Add support for printing slices via print/println

### LINKED ISSUE #2922: print() complains "unknown arg type: []byte"
Trying to use a byte array throws a compilation error : 

unknown arg type: []byte

Doesnt matter if self written or in any predesigned struct like bytes.buffer{}

tinygo version 0.23.0 windows/amd64 (using go version go1.18.3 and LLVM version 14.0.0)
MCU used: RPI PICO

### PR DESCRIPTION
**Pull request by fgsch**
Fixes #2922.

### CODE REVIEW

File: compiler/compiler.go [hunk 1]
```diff
@@ -1519,6 +1519,8 @@ func (b *builder) createBuiltin(argTypes []types.Type, argValues []llvm.Value, c
 			case *types.Pointer:
 				ptrValue := b.CreatePtrToInt(value, b.uintptrType, "")
 				b.createRuntimeCall("printptr", []llvm.Value{ptrValue}, "")
+			case *types.Slice:
+				b.createRuntimeCall("printslice", []llvm.Value{value}, "")
```
Comment (aykevl) [inline compiler/compiler.go]: You need to split up the slice in pointer/len/cap and cast the pointer to a consistent type (like unsafe.Pointer).

Comment (fgsch) [reply compiler/compiler.go]: I had that originally but this works the same. Is there any reason to use one over the other?

Comment (aykevl) [reply compiler/compiler.go]: Yes, try printing a slice that's not a byte slice. []int for example.

File: testdata/print.go [hunk 2]
```diff
@@ -45,4 +45,7 @@ func main() {
 
 	// print bool
 	println(true, false)
+
+	// print slice
+	// println([]byte{'H', 'e', 'l', 'l', 'o', ' ', 'W', 'o', 'r', 'l', 'd'})
```
Comment (aykevl) [inline testdata/print.go]: I recommend printing `[]byte(nil)` and `[]int(nil)`. Not a full test but using a non-nil slice is not possible.

Comment (fgsch) [reply testdata/print.go]: Good idea. I will update it.

File: compiler/compiler.go (hunk 1 above)
Comment (fgsch) [reply compiler/compiler.go]: After I updated the tests with your suggestion I noticed this 😅

Comment (fgsch) [reply compiler/compiler.go]: Done.

File: testdata/print.go (hunk 2 above)
Comment (fgsch) [reply testdata/print.go]: Done.

File: src/runtime/print.go [hunk 3]
```diff
@@ -372,3 +372,12 @@ func printbool(b bool) {
 		printstring("false")
 	}
 }
+
+func printslice(ptr uintptr, len_, cap_ int64) {
```
Comment (aykevl) [inline src/runtime/print.go]: `len` and `cap` are also of `uintptr` type. They just happen to match `int64` on a 64-bit system.

File: src/runtime/print.go [hunk 4]
```diff
@@ -349,7 +349,7 @@ func printmap(m *hashmap) {
 
 func printptr(ptr uintptr) {
 	if ptr == 0 {
-		print("nil")
+		print("0x0")
```
Comment (aykevl) [inline src/runtime/print.go]: Any reason for this change? I know it didn't match upstream exactly, but `nil` seems fine to me.

Comment (fgsch) [reply src/runtime/print.go]: To match upstream. I will revert it.

File: src/runtime/print.go (hunk 3 above)
Comment (fgsch) [reply src/runtime/print.go]: I think I can use `int` in this case too to avoid converting it back? That should work in both 32 and 64 bit systems.

Comment (aykevl) [reply src/runtime/print.go]: There are also 16-bit systems: AVR. While `len(slice)` returns an int on AVR to comply with the spec, it is internally stored as a pointer-sized integer (which is 16 bits on AVR) and thus matches `uintptr`.
You can add a new `printuintptr` like this:

```go
func printuintptr(n uintptr) {
    switch unsafe.Sizeof(n) {
    case 2:
        printuint16(uint16(n))
    // etc
    }
}
```

You can test whether it works there by running `tinygo run -target=simavr ./testdata/print.go`.

Comment (fgsch) [reply src/runtime/print.go]: Thanks again for the cluebat. I will update it.

Comment (fgsch) [reply src/runtime/print.go]: I used `print(uint(len_))` and `print(uint(cap_))` instead. Hopefully this is good.

**Comment by dkegel-fastly** [comment]
Try this:
```
--- a/.github/workflows/linux.yml
+++ b/.github/workflows/linux.yml
@@ -261,6 +261,7 @@ jobs:
         uses: actions/checkout@v2
       - name: Install apt dependencies
         run: |
+          sudo apt-get update
           sudo apt-get install --no-install-recommends \
               qemu-user \
               g++-arm-linux-gnueabihf \
```

**Comment by fgsch** [comment]
@aykevl updated and tested in avr. Is there anything else I should address before this can be merged?

File: src/runtime/print.go (hunk 3 above)
Comment (aykevl) [reply src/runtime/print.go]: This works, but it results in larger code than necessary on AVR. (Note that the switch on `unsafe.Sizeof` is resolved at compile time, not at runtime). So I'd really prefer a `printuintptr` that prints numbers based on the actual integer size.

Here is some code that also does this, and can be factored out:

https://github.com/tinygo-org/tinygo/blob/b65447c7d567eea495805656f45472cc3c483e03/src/runtime/print.go#L326-L333

Comment (fgsch) [reply src/runtime/print.go]: Ok, I will update it accordingly.

Comment (fgsch) [reply src/runtime/print.go]: Done

**Comment by deadprogram** [comment]
Seems to me this is ready, once the commits are squashed, if you please. Or would you prefer I do that for you?

**Comment by fgsch** [comment]
I will do it shortly. Thank you!

**Comment by fgsch** [comment]
@deadprogram done.

**Comment by deadprogram** [comment]
Thanks for implementing this @fgsch now merging.

**Pull request merged**